experimental/InvariantTheory/src/InvariantTheory.jl

```diff
@@ -0,0 +1,654 @@
+null_cone_ideal(R::InvariantRing) = R.NullConeIdeal
+
+function Base.show(io::IO, R::InvariantRing) #TODO compact printing
+    #if get(io, :supercompact, false)
+        print(io, "Invariant Ring of", "\n")
+        show(io, R.poly_ring)
+        print(io, " under group action of ", R.group.group[1], R.group.group[2], "\n", "\n")
+        print(io, "Generated by ", "\n")
+    join(io, R.fundamental, "\n")
+    #else
+     #   print(io, "Invariant Ring under ")
+      #  show(io, R.group[1], R.group[2])
+    #end
+end
+
+#I think this should work for everyone except tensor products.
+function image_ideal(G::ReductiveGroup, rep_mat::AbstractAlgebra.Generic.MatSpaceElem)
+    R = base_ring(rep_mat)
+    n = ncols(rep_mat)
+    m = G.group[2]
+    #r = G.direct_product[2]
+    mixed_ring_xy, x, y, zz = PolynomialRing(G.field, "x"=>1:n, "y"=>1:n, "zz"=>1:m^2)
+    # naming the variables zz here and z in the group ring
+    #for determinant - 
+    #M1 = matrix(mixed_ring_xy,m,m,[zz[i,j] for i in 1:m for j in 1:m])
+    ztozz = hom(R,mixed_ring_xy, gens(mixed_ring_xy)[(2*n)+1:(2*n)+(m^2)])
+    genss_ = gens(G.group_ideal)
+    genss = [ztozz(genss_[i]) for i in 1:length(genss_)]
+    #rep_mat in the new ring
+    new_rep_mat = matrix(mixed_ring_xy,n,n,[ztozz(rep_mat[i,j]) for i in 1:n, j in 1:n])
+    new_vars = new_rep_mat*[x[i] for i in 1:n] 
+    ideal_vect = [y[i] - new_vars[i] for i in 1:n] 
+    ideal_vect = vcat(ideal_vect,genss)
+    return ideal(mixed_ring_xy, ideal_vect), new_rep_mat
+end
+
+#tested
+function factorise(x::MPolyRingElem)
+    #x has to be a monomial. No check implemented. 
+    R = parent(x)
+    V = exponent_vector(x,1)
+    Factorisation = Vector{Tuple{elem_type(R), Int}}()
+    for i in 1:length(V)
+        push!(Factorisation, (gens(R)[i], V[i]))
+    end
+    return Factorisation
+end
+
+
+function proj_of_image_ideal(G::ReductiveGroup, rep_mat::AbstractAlgebra.Generic.MatSpaceElem)
+    W = image_ideal(G, rep_mat)
+    mixed_ring_xy = base_ring(W[2])
+    n = ncols(rep_mat)
+    m = G.group[2]
+    #r = G.direct_product[2]
+    #use parallelised groebner bases here. This is the bottleneck!
+    return eliminate(W[1], gens(mixed_ring_xy)[(2*n)+1:(2*n)+(m^2)]), W[2]
+end
+
+
+#evaluate at y = 0 
+function generators(G::ReductiveGroup, X::MPolyIdeal, rep_mat::AbstractAlgebra.Generic.MatSpaceElem)
+    n = ncols(rep_mat)
+    m = G.group[2]
+    gbasis = gens(X) 
+    length(gbasis) == 0 && return gbasis,new_rep_mat
+    mixed_ring_xy = parent(gbasis[1])
+    
+    #to evaluate gbasis at y = 0
+    ev_gbasis = Vector{Union{AbstractAlgebra.Generic.MPoly{nf_elem}, MPolyRingElem}}(undef,0)
+    for elem in gbasis
+        b = mixed_ring_xy()
+        mons = collect(monomials(elem))
+        coeffs = collect(coefficients(elem))
+        for i in 1:length(mons)
+            if (exponent_vector(mons[i],1)[n+1:2*n] == [0 for i in 1:n])
+                b += coeffs[i]*mons[i]
```

Comment:
Couldn't you just iterate over `terms` instead? Like this perhaps:
```suggestion
        for t in terms(elem)
            if exponent_vector(t,1)[n+1:2*n] == [0 for i in 1:n]
                b += t
```